networkit/cpp/distance/Diameter.cpp

```diff
@@ -52,38 +65,124 @@ std::pair<count, count> Diameter::getDiameter() const {
     return diameterBounds;
 }
 
+count Diameter::getNumBFS() const {
+  return numBFS;
+}
+
+count Diameter::getInitialLB() const {
+  return initialLB;
+}
+
 edgeweight Diameter::exactDiameter(const Graph& G) {
     using namespace std;
 
     Aux::SignalHandler handler;
 
     edgeweight diameter = 0.0;
 
-    if (! G.isWeighted()) {
-        std::tie(diameter, std::ignore) = estimatedDiameterRange(G, 0);
-    } else {
-        G.forNodes([&](node v) {
-            handler.assureRunning();
-            Dijkstra dijkstra(G, v);
-            dijkstra.run();
-            auto distances = dijkstra.getDistances();
-            G.forNodes([&](node u) {
-                if (diameter < distances[u]) {
-                    diameter = distances[u];
-                }
-            });
+    G.forNodes([&](node v) {
+        handler.assureRunning();
+        std::unique_ptr<SSSP> sssp;
+        if (G.isWeighted()) {
+            sssp = std::make_unique<Dijkstra>(G, v);
+        } else {
+            sssp = std::make_unique<BFS>(G, v);
+        }
+        sssp->run();
+        const auto &distances = sssp->getDistances();
+        G.forNodes([&](node u) {
+            if (diameter < distances[u]) {
+                diameter = distances[u];
+            }
         });
-    }
+    });
 
     if (diameter == std::numeric_limits<edgeweight>::max()) {
         throw std::runtime_error("Graph not connected - diameter is infinite");
     }
     return diameter;
 }
 
+std::pair<edgeweight, edgeweight> Diameter::difub(const Graph &G, node u, double error) {
+    Aux::SignalHandler handler;
+
+    handler.assureRunning();
+    std::vector<std::vector<count>> distancesF(1);
+    std::vector<std::vector<count>> distancesB(1);
+
+    // calculate all forward and backward distances
+    handler.assureRunning();
+    Traversal::BFSfrom(G, u, [&](node v, count dist) {
+        if (distancesF.size() <= dist) {
+            distancesF.emplace_back();
+        }
+        distancesF[dist].push_back(v);
+    });
+    Traversal::BFSfrom(G, u, [&](node v, count dist) {
+        if (distancesB.size() <= dist) {
+            distancesB.emplace_back();
+        }
+        distancesB[dist].push_back(v);
+    }, true);
+
+    count i = std::max(distancesF.size(), distancesB.size());
+    initialLB = i;
+    std::atomic<count> lb(i);
+    count ub = 2 * i;
+
+    numBFS = 2;
+
+    // difub
+    for (; i > 0; --i) {
+        handler.assureRunning();
+        count size_f = i < distancesF.size() ? distancesF[i].size() : 0;
+        count size_b = i < distancesB.size() ? distancesB[i].size() : 0;
+#pragma omp parallel for
+        for (count j = 0; j < size_f + size_b; ++j) {
```

Comment:
This needs to be `omp_index` which is a signed integer when using MSVC as - at lest with the parameters NetworKit uses - MSVC doesn't support OpenMP loops with unsigned integers.

Reply:
Thanks, I fixed it. Changing `j` to `omp_index` should be enough? In `cpp/distance/APSP.cpp` the loop has a `count` in its condition.